engine/src/engine.rs

```diff
@@ -1650,13 +1653,43 @@ impl<'env, J: IO + Copy, E: Env> ApplyBackend for Engine<'env, J, E> {
                     }
                 }
                 Apply::Delete { address } => {
+                    let current_basic = self.basic(address);
+                    accounting.remove(current_basic.balance);
+
                     let address = Address::new(address);
                     let generation = get_generation(&self.io, &address);
                     remove_account(&mut self.io, &address, generation);
                     writes_counter += 1;
                 }
             }
         }
+        match accounting.net() {
+            // Net loss is possible if `SELFDESTRUCT(self)` calls are made.
+            accounting::Net::Lost(amount) => {
+                sdk::log!(
+                    crate::prelude::format!("Burn {} ETH due to SELFDESTRUCT", amount).as_str()
+                );
+                // Apply changes for eth-connector. We ignore the `StorageReadError` intentionally since
+                // if we cannot read the storage then there is nothing to remove.
+                EthConnectorContract::init_instance(self.io)
+                    .map(|mut connector| {
+                        // The `unwrap` is safe here because (a) if the connector
+                        // is implemented correctly then the total supply will never underflow and (b) we are passing
+                        // in the balance directly so there will always be enough balance.
+                        connector.internal_remove_eth(Wei::new(amount)).unwrap();
+                    })
+                    .ok();
+            }
+            accounting::Net::Zero => (),
+            accounting::Net::Gained(_) => {
+                // It should be impossible to gain ETH using normal EVM operations in production.
+                // In tests we have convenience functions that can poof addresses with ETH out of nowhere.
+                #[cfg(all(not(feature = "integration-test"), feature = "contract"))]
+                {
+                    unreachable!()
```

Comment:
I get that this is unreachable but I would still include it as an error so that if someone somehow gets this, then they can at least refer us to an `ERR_INFLATION` or something similar error.

Reply:
Done in https://github.com/aurora-is-near/aurora-engine/pull/536/commits/de999696015d6fdc8f9a394f15295398bfc1f819